cmd/krel/cmd/changelog.go

```diff
@@ -109,56 +111,79 @@ func init() {
 }
 
 func runChangelog() (err error) {
-	// TODO: remote lookup of 'final release notes draft',
-	// including error handling if not found
+	tag, err := semver.Make(util.TrimTagPrefix(changelogOpts.tag))
+	if err != nil {
+		return err
+	}
+	branch := fmt.Sprintf("release-%d.%d", tag.Major, tag.Minor)
+	logrus.Infof("Using release branch %s", branch)
 
-	markdown, version, err := createReleaseNotes()
+	repo, err := git.CloneOrOpenDefaultGitHubRepoSSH(rootOpts.repoPath, git.DefaultGithubOrg)
 	if err != nil {
 		return err
 	}
+
+	var markdown string
+
+	if tag.Patch == 0 {
+		if len(tag.Pre) == 0 {
+			// New final minor versions should have remote release notes
+			markdown, err = lookupRemoteReleaseNotes(branch)
+		} else {
+			// New minor alphas, betas and rc get generated notes
+			start, e := repo.PreviousTag(changelogOpts.tag, branch)
+			if e != nil {
+				return e
+			}
+			logrus.Infof("Found previous tag %s", start)
+			markdown, err = generateReleaseNotes(branch, start, changelogOpts.tag)
+		}
+	} else {
+		// A patch version, let’s just use the previous patch
+		start := util.AddTagPrefix(semver.Version{
+			Major: tag.Major, Minor: tag.Minor, Patch: tag.Patch - 1,
+		}.String())
+
+		markdown, err = generateReleaseNotes(branch, start, changelogOpts.tag)
+	}
+	if err != nil {
+		return err
+	}
+
 	toc, err := notes.GenerateTOC(markdown)
 	if err != nil {
 		return err
 	}
 
-	if err := writeMarkdown(toc, markdown, version); err != nil {
+	if err := writeMarkdown(toc, markdown, tag); err != nil {
 		return err
 	}
 
-	if err := writeHTML(version, markdown); err != nil {
+	if err := writeHTML(tag, markdown); err != nil {
 		return err
 	}
 
 	// TODO: Push changes into repo
 	return nil
 }
 
-func createReleaseNotes() (markdown, version string, err error) {
-	revisionDiscoveryMode := notes.RevisionDiscoveryModeMinorToMinor
-
-	if changelogOpts.branch != git.Master {
-		if !git.IsReleaseBranch(changelogOpts.branch) {
-			return "", "", errors.Wrapf(
-				err, "Branch %q is no release branch", changelogOpts.branch,
-			)
-		}
-		revisionDiscoveryMode = notes.RevisionDiscoveryModePatchToPatch
-	}
-	logrus.Infof("Using branch %q", changelogOpts.branch)
-	logrus.Infof("Using discovery mode %q", revisionDiscoveryMode)
+func generateReleaseNotes(branch, startRev, endRev string) (string, error) {
+	logrus.Info("Generating release notes")
 
 	notesOptions := notes.NewOptions()
-	notesOptions.Branch = changelogOpts.branch
-	notesOptions.DiscoverMode = revisionDiscoveryMode
+	notesOptions.Branch = branch
+	notesOptions.StartRev = startRev
+	notesOptions.EndRev = endRev
```

Comment:
Are these required for `notesOptions.ValidateAndFinish()`? Wondering if some of these options being set in this list could either be passed as parameters to `NewOptions()` or configured as sane defaults. What do you think?

Reply:
Yes exactly. Yeah we could apply something like a builder pattern for validation... :thinking:

Reply:
That is exactly what I was thinking :) Given that this does not change functionality, would you prefer for this to be tackled in a follow-up PR?

Reply:
I would like to defer that to a follow-up PR if you don't mind :)

Reply:
SGTM!